### PR TITLE
Fanout event source swallows errors

### LINKED ISSUE #6741: Fan out event source not handling errors and blocking process exit when no events
Fan out event source not handling errors and blocking process exit when no events.

### PR DESCRIPTION
**Pull request by ettec**
closes #6741

### CODE REVIEW

**Review by pscott31 — DISMISSED** [review]

**Review by zale144 — DISMISSED** [review]

File: datanode/broker/buffered_event_source.go [hunk 1]
```diff
@@ -106,7 +108,10 @@ func (m *FileBufferedEventSource) writeEventsToBuffer(ctx context.Context, sourc
 	var err error
 	for {
 		select {
-		case event := <-sourceEventCh:
+		case event, ok := <-sourceEventCh:
+			if !ok {
```
[the remark this answers is not in the document]
Comment (ettec) [reply datanode/broker/buffered_event_source.go:112]: channel closure is not unexpected, jsut need to handle it

File: datanode/broker/buffered_event_source.go [hunk 2]
```diff
@@ -138,8 +143,11 @@ func (m *FileBufferedEventSource) writeEventsToBuffer(ctx context.Context, sourc
 				m.lastBufferedSeqNum <- bufferSeqNum
 			}
 
-		case err = <-sourceErrCh:
-			sinkErrorCh <- err
+		case srcErr, ok := <-sourceErrCh:
+			if !ok {
```
[the remark this answers is not in the document]
Comment (ettec) [reply datanode/broker/buffered_event_source.go:147]: same here, channel closure is not unexpected, just need to handle it - these things are chained so the upstream broker closing its channels is normal behaviour

File: datanode/broker/buffered_event_source.go [hunk 3]
```diff
@@ -77,12 +77,14 @@ func (m *FileBufferedEventSource) Receive(ctx context.Context) (<-chan events.Ev
 	sinkEventCh := make(chan events.Event, m.sendChannelBufferSize)
 	sinkErrorCh := make(chan error, 1)
 
+	ctxWithCancel, cancelFunc := context.WithCancel(ctx)
```
[the remark this answers is not in the document]
Comment (ettec) [reply datanode/broker/buffered_event_source.go:80]: the different context name is to differentiate the one passed into the read buffer from the one passed into write buffer.  The one passed into read buffer is cancelled when the write buffer exits.

**Review by jeremyletang — CHANGES_REQUESTED** [review]
Seems to be blocking for me locally, might need some more testing.

Can you try on the branch instead of the cherry pick, I can't reproduce and looking at the change I'm not sure how it could cause a block?

File: datanode/broker/buffered_event_source.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ettec) [reply datanode/broker/buffered_event_source.go]: sorted the cancelFunc name

**Comment by zale144** [comment]
lgtm! 
@vegaprotocol/core 
Because the `broker.Receive` doesn't handle context cancel, and in order to prevent the `receiveCh` channel range from blocking, I would just consider adding this change to `broker.go`:

```
func (b *Broker) Receive(ctx context.Context) error {
	if err := b.eventSource.Listen(); err != nil {
		return err
	}

	receiveCh, errCh := b.eventSource.Receive(ctx)

	for {
		select {
		case <-ctx.Done():
			return nil
		case err := <-errCh:
			return err
		case e, ok := <-receiveCh:
			if !ok {
				return nil
			}
			if err := checkChainID(b.chainID, e.ChainID()); err != nil {
				return err
			}
			go b.Send(e)
		}
	}
}
```

Not sure if the `go b.Send(e)` is a good idea. Thoughts?

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**